keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,421 @@
+
+### User Stories [optional]
+
+#### Story 1
+As a cluster admin, I would like my node to be as performant as possible, and not waste resources by duplicating stats collection by CRI and cAdvisor.
+
+#### Story 2
+As a node maintainer, I would like to clarify the sources of truth for each stat the Kubelet reports to the Summary API, so that I can better find which component is to investigate for any issues.
+
+### A note on history
+There have been conversations in the past on [deprecating the stats Summary API][#core-proposal], in favor of allowing an on-demand model of stats gathering/reporting.
+This KEP does not attempt to implement that. These stats are largely considered [a part of Kubernetes][#keep-summary]. The stated goal of this KEP is to reduce the overhead
+on two entities reporting metrics, not totally changing what stats the Kubelet reports (with the exception of Accelerator and UserDefinedMetrics).
+
+Thus, this KEP largely the plan described [here][#plan], with some changes:
+
+- the `/metrics/cadvisor` endpoint will likely not be exposed
+	- **TODO** How to prevent this from breaking users (feature gate?)
+- CRI is used for all of the monitoring endpoints related to Containers and Pods (except Volume and Ephemeral Storage)
+- CRI is used to provide metrics for eviction.
+
+[core-proposal]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/core-metrics-pipeline.md#proposed-core-metrics
+[keep-summary]: https://github.com/kubernetes/kubernetes/issues/68522#issuecomment-666636130
+[plan]: https://github.com/kubernetes/kubernetes/issues/68522#issuecomment-724928827
+### Risks and Mitigations
+
+- To properly move to CRI stats, it is likely there are some metrics that we'll want to not support (Accelerator/UserDefined). We should be careful to not break entities that rely on these metrics.
+- A large part of this work is changing the source of the Summary API metrics. In doing so, there is a risk that collecting from a new source will change how the stats look in aggregate (and risk bugs popping up in new areas).
+- cAdvisor has a long history of collecting these stats. There is a risk that changing the source of the stats to the CRI implementation can cause performance regressions
+as cAdvisor is fine tuned to perform in an adequate manner.
+    - CRI implementations should do performance regression analyses to ensure the change does not regress too much.
+
+## Graduation Criteria
+
+
+## Implementation History
+
+
+## Drawbacks [optional]
+
+CRI runtimes will each have to implement additional interface to support full stats, rather than all metric collection being unified by cAdvisor. 
+
+Note: This is by design as this will enable to decouple runtime implementation details further from kubelet.
+
+## Alternatives [optional]
```

Comment:
One other alternative to consider:

Since the summary API and /metrics/cadvisor carry roughly equivalent sets of data, we could reimplement /metrics/cadvisor using data from the CRI.  There may be some small inconsistencies, but it may be acceptable to users.

Reply:
I honestly didn't know about it until I was pointed at that issue you posted about deprecating JSON. Do you have a pointer to the API definition? In the new version, we could translate between CRI and /metrics/cadvisor and have kubelet serve both that and /stats/summary

Reply:
https://github.com/google/cadvisor/tree/master/metrics contains the code to generate the prometheus metrics.  The endpoint is commonly used by prometheus users.  See https://github.com/prometheus/prometheus/blob/67edfc6054ec07cef8b4f5faed227c5ddcf4d0fe/documentation/examples/prometheus-kubernetes.yml#L103 for the example config to scrape cAdvisor's metrics.

Reply:
not only prometheus, but also `kubectl top`..